kep/399-sysched-scoring/README.md

```diff
@@ -0,0 +1,217 @@
+
+<p align="center">
+<img src="images/example_complete.png" style="width:50%">
+</p>
+
+<p align = "center">
+<b>Figure 3. Scheduling algorithm and pod placement example</b>
+</p>
+
+Figure 3 also highlights the opportunities not only for reducing the exposure to extraneous system calls (i.e., ExS score) but also the node’s attack surface. If $P_3$ had been scheduled on Node 2 instead of Node 1, then Node 2 would have a total of 7 system calls "open" $(1,3,4,5,7,8,9)$, instead of only 3 system calls under the SySched approach. 
+
+### Scheduling Plugin
+
+There are two main components of our plugin: 1) a lightweight in-memory cache that is kept synchronize with pod events in order to keep an up-to-date node to pod mapping in the cluster, and 2) mechanisms for retrieving the system call sets used by pods for computing the ExS scores. The in-memory cache reduces the cost of querying the API server for each pod scheduling event. We obtain the system call profiles for pods from their seccomp profiles. Our plugin leverages the Kubernetes sigs community-developed [Security Profiles Operator](https://github.com/kubernetes-sigs/security-profiles-operator) (SPO) that creates Custom Resource Definitions (CRDs) for seccomp profiles. If a pod has no seccomp profile associated with it, then our plugin simply returns, effectively a no-op.
+
+#### **How the Scheduler gets access to pods' system call profiles**
+
+We assume that the system call profiles of pods have explicitly been made available to the scheduler via the SPO. We rely on the SPO to generate and bind seccomp profile CRDs to the pods. The scheduler uses the API server to access a pod's seccomp profile CRD and parses the seccomp profile CRD to obtain the list of system calls for a pod. 
+
+An example of an SPO's seccomp profile CRD:
+
+```
+---
+apiVersion: security-profiles-operator.x-k8s.io/v1beta1
+kind: SeccompProfile
+metadata:
+  name: nginx-seccomp
+  namespace: scheduler-plugins
+spec:
+  defaultAction: SCMP_ACT_LOG
+  architectures:
+  - SCMP_ARCH_X86_64
+  syscalls:
+  - action: SCMP_ACT_ALLOW
+    names:
+    - recvmsg
+    - fchdir
+    - sendfile
+    - getdents64
+    - setsockopt
+    - recvfrom
+    - sigaltstack
+    - fstatfs
+    - umask
+    - getegid
+    - read
```

Comment:
I'm wondering if we wanna create groups of syscalls depending on their security impact. For example the `RuntimeDefault` profile already tries to provide strong defaults.

Reply:
Grouping syscalls based on their security impact is definitely something interesting. We thought about that aspect when we mentioned the use of weights to denote the "criticality" of certain syscalls. We can define criticality in a few ways but one way is to base on historical CVE data associated with syscalls or how syscalls are used in known exploits. 

Introducing this groupings/weights through the scheduling plugin is one way or encode it somehow in the SPO CRD would also be useful.